File: conv.go
package nrsc5

import "fmt"

// Viterbi decoder for the NRSC-5 convolutional codes, ported from
// conv_dec.c / conv_gen.h (originally from Ettus Research, GPLv3).

// Termination types
const (
	convTermFlush       = iota // zero flush termination
	convTermTailBitting        // tail biting termination
)

const tailBitingExtra = 32

// lteConvCode describes a convolutional code.
//
//	n    - Rate 2, 3, 4 (1/2, 1/3, 1/4)
//	k    - Constraint length (7 or 9)
//	len  - Number of information bits
//	rgen - Recursive generator polynomial in octal (0 for non-recursive)
//	gen  - Generator polynomials in octal
//	term - Termination type
type lteConvCode struct {
	n    int
	k    int
	len  int
	rgen uint
	gen  [4]uint
	term int
}

// convDecode performs Viterbi decoding of a tail-biting convolutional code.
// in contains 3 soft bits (NRZ-mapped: >0 is 1, <0 is 0) per input bit;
// out receives len decoded bits (0 or 1).
// Returns the difference between the two best final path metrics
// (a confidence measure, not an error count).
func convDecode(in []int8, out []uint8, code lteConvCode) (int, error) {
	k := code.k
	var ns int
	switch k {
	case 7:
		ns = 64
	case 9:
		ns = 256
	default:
		return 0, fmt.Errorf("nrsc5: unsupported constraint length %d", k)
	}
	if code.n != 3 {
		return 0, fmt.Errorf("nrsc5: unsupported code rate 1/%d", code.n)
	}

	// Tail biting decoders run through the trellis an extra
	// TAIL_BITING_EXTRA cycles on each end.
	decLen := code.len
	if code.term != convTermTailBitting {
		decLen += k - 1
	} else {
		decLen += 2 * tailBitingExtra
	}

	// Trellis outputs (NRZ-mapped) and input bit per state.
	outputs := make([]int16, ns*4)
	vals := make([]uint8, ns)
	for state := 0; state < ns; state++ {
		var o [4]int16
		if code.rgen != 0 {
			genRecStateInfo(code, &vals[state], uint(state), o[:])
		} else {
			genStateInfo(code, &vals[state], uint(state), o[:])
		}
		copy(outputs[4*state:4*state+4], o[:])
	}

	// Sums is the accumulated path metric per state.
	sums := make([]int16, ns)

	// Paths holds the butterfly decision for each time step.
	paths := make([][]int16, decLen)
	paths[0] = make([]int16, ns*decLen)
	for i := 1; i < decLen; i++ {
		paths[i] = paths[0][i*ns : (i+1)*ns]
	}

	// Normalization interval.
	intrvl := (1<<15)/(code.n*127) - k

	// Propagate through the trellis.
	j := 0
	if code.term == convTermTailBitting {
		j = code.len - tailBitingExtra
	}
	for i := 0; i < decLen; i, j = i+1, j+1 {
		if code.term == convTermTailBitting && j == code.len {
			j = 0
		}
		seq := in[3*j : 3*j+3]
		metrics := make([]int16, ns/2)
		for s := 0; s < ns/2; s++ {
			metrics[s] = int16(seq[0])*outputs[4*s+0] +
				int16(seq[1])*outputs[4*s+1] +
				int16(seq[2])*outputs[4*s+2]
		}

		// ACS butterfly for states 0..ns/2-1. State pairs (2s, 2s+1)
		// transition to states s and s+ns/2.
		newSums := make([]int16, ns)
		for s := 0; s < ns/2; s++ {
			state0 := sums[2*s]
			state1 := sums[2*s+1]

			sum0 := int(state0) + int(metrics[s])
			sum1 := int(state1) - int(metrics[s])
			sum2 := int(state0) - int(metrics[s])
			sum3 := int(state1) + int(metrics[s])

			if sum0 > sum1 {
				newSums[s] = int16(sum0)
				paths[i][s] = -1
			} else {
				newSums[s] = int16(sum1)
				paths[i][s] = 0
			}

			if sum2 > sum3 {
				newSums[s+ns/2] = int16(sum2)
				paths[i][s+ns/2] = -1
			} else {
				newSums[s+ns/2] = int16(sum3)
				paths[i][s+ns/2] = 0
			}
		}

		if i%intrvl == 0 {
			min := newSums[0]
			for i2 := 1; i2 < ns; i2++ {
				if newSums[i2] < min {
					min = newSums[i2]
				}
			}
			for i2 := 0; i2 < ns; i2++ {
				newSums[i2] -= min
			}
		}

		copy(sums, newSums)
	}

	// Traceback.
	state := 0
	var maxVal, maxPrev int
	if code.term == convTermTailBitting {
		maxVal, maxPrev = -(1 << 30), -(1 << 30)
		for i := 0; i < ns; i++ {
			s := int(sums[i])
			if s > maxVal {
				maxPrev = maxVal
				maxVal = s
				state = i
			}
		}
		if maxVal < 0 {
			return 0, fmt.Errorf("nrsc5: trellis overflow")
		}
		// Discard the extra TAIL_BITING_EXTRA transitions at the end.
		for i := decLen - 1; i >= code.len+tailBitingExtra; i-- {
			path := int(paths[i][state]) + 1
			state = int(vstateLshift(uint(state), k, path))
		}
	} else {
		for i := decLen - 1; i >= code.len; i-- {
			path := int(paths[i][state]) + 1
			state = int(vstateLshift(uint(state), k, path))
		}
	}

	// Traceback through the data portion, writing decoded bits.
	off := 0
	if code.term == convTermTailBitting {
		off = tailBitingExtra
	}
	if code.rgen != 0 {
		for i := code.len - 1; i >= 0; i-- {
			path := int(paths[i+off][state]) + 1
			out[i] = uint8(path) ^ vals[state]
			state = int(vstateLshift(uint(state), k, path))
		}
	} else {
		for i := code.len - 1; i >= 0; i-- {
			path := int(paths[i+off][state]) + 1
			out[i] = vals[state]
			state = int(vstateLshift(uint(state), k, path))
		}
	}

	return maxVal - maxPrev, nil
}

// vstateLshift left shifts a state register and inserts val as LSB.
func vstateLshift(reg uint, k, val int) uint {
	var mask uint
	switch k {
	case 5:
		mask = 0x0e
	case 7:
		mask = 0x3e
	case 9:
		mask = 0xfe
	}
	return ((reg << 1) & mask) | uint(val)
}

// genStateInfo populates non-recursive trellis state info.
func genStateInfo(code lteConvCode, val *uint8, reg uint, out []int16) {
	// Previous '0' state
	prev := vstateLshift(reg, code.k, 0)

	// Compute output and unpack to NRZ
	*val = uint8((reg >> (code.k - 2)) & 0x01)
	prev |= uint(*val) << (code.k - 1)

	for i := 0; i < code.n; i++ {
		out[i] = int16(parity(prev&code.gen[i])*2 - 1)
	}
}

// genRecStateInfo populates recursive trellis state info.
func genRecStateInfo(code lteConvCode, val *uint8, reg uint, out []int16) {
	// Previous '0' state
	prev := vstateLshift(reg, code.k, 0)

	// Compute recursive input value
	rec := (reg >> (code.k - 2)) & 0x01

	if parity(prev&code.rgen) == int(rec) {
		*val = 0
	} else {
		*val = 1
	}

	// Compute outputs and unpack to NRZ
	prev |= rec << (code.k - 1)

	var mask uint
	switch code.k {
	case 5:
		mask = 0x0f
	case 7:
		mask = 0x3f
	default:
		mask = 0xff
	}

	for i := 0; i < code.n; i++ {
		if code.gen[i]&mask != 0 {
			out[i] = int16(parity(prev&code.gen[i])*2 - 1)
		} else {
			out[i] = int16(*val*2 - 1)
		}
	}
}

func parity(x uint) int {
	x ^= x >> 32
	x ^= x >> 16
	x ^= x >> 8
	x ^= x >> 4
	x ^= x >> 2
	x ^= x >> 1
	return int(x & 1)
}

// Convolutional code parameters used by NRSC-5.
var (
	convCodeK7   = lteConvCode{n: 3, k: 7, gen: [4]uint{0133, 0171, 0165}, term: convTermTailBitting}
	convCodeE1   = lteConvCode{n: 3, k: 9, gen: [4]uint{0561, 0657, 0711}, term: convTermTailBitting}
	convCodeE2E3 = lteConvCode{n: 3, k: 9, gen: [4]uint{0561, 0753, 0711}, term: convTermTailBitting}
)

// ConvDecodeP1 decodes the FM P1 frame (rate 1/3, K=7, tail biting).
func ConvDecodeP1(in []int8, out []uint8) (int, error) {
	code := convCodeK7
	code.len = P1_FRAME_LEN_FM
	return convDecode(in, out, code)
}

// ConvDecodePIDS decodes the FM PIDS frame (rate 1/3, K=7, tail biting).
func ConvDecodePIDS(in []int8, out []uint8) (int, error) {
	code := convCodeK7
	code.len = PIDS_FRAME_LEN
	return convDecode(in, out, code)
}

// ConvDecodeP3P4 decodes FM P3/P4 frames (rate 1/3, K=7, tail biting).
func ConvDecodeP3P4(in []int8, out []uint8, length int) (int, error) {
	code := convCodeK7
	code.len = length
	return convDecode(in, out, code)
}

// ConvDecodeE1 decodes AM P1 frames (rate 1/3, K=9, tail biting).
func ConvDecodeE1(in []int8, out []uint8, length int) (int, error) {
	code := convCodeE1
	code.len = length
	return convDecode(in, out, code)
}

// ConvDecodeE2E3 decodes AM PIDS/P3 frames (rate 1/3, K=9, tail biting).
func ConvDecodeE2E3(in []int8, out []uint8, length int) (int, error) {
	code := convCodeE2E3
	code.len = length
	return convDecode(in, out, code)
}

File: python/nrsc5py/conv.py
"""Viterbi decoder for the NRSC-5 convolutional codes (mirrors conv.go).

The NRSC-5 L1 channels use rate-1/3 convolutional codes with
tail-biting termination:

    P1 / PIDS / P3 (FM)   K=7, generators 0133, 0171, 0165 (octal)
    AM E1                 K=9, generators 0561, 0657, 0711
    AM E2/E3              K=9, generators 0561, 0753, 0711

Tail-biting means the encoder's shift register is initialized with the
last K-1 data bits, so the trellis starts and ends in the same state.
The decoder runs the trellis an extra 32 steps on each end (a "wrap"
period) and picks the best final state, then traces back.

The heavy loop (Add-Compare-Select over all states for every input bit)
is JIT-compiled with numba. Everything else stays readable Python.
"""

from __future__ import annotations

import numpy as np

try:
    from numba import njit
except ImportError:  # pragma: no cover - fall back to slow path
    def njit(*args, **kwargs):
        if len(args) == 1 and callable(args[0]):
            return args[0]
        def wrap(fn):
            return fn
        return wrap

from .defines import P1_FRAME_LEN_FM, PIDS_FRAME_LEN

# Termination types.
CONV_TERM_FLUSH = 0
CONV_TERM_TAIL_BITING = 1

# Extra trellis steps at each end for tail-biting codes.
TAIL_BITING_EXTRA = 32

# Generator polynomials in octal, matching the C code's LTE-style
# descriptors. gen[i] taps the shift register to produce output i.
CONV_CODE_K7 = (7, (0o133, 0o171, 0o165))
CONV_CODE_E1 = (9, (0o561, 0o657, 0o711))
CONV_CODE_E2_E3 = (9, (0o561, 0o753, 0o711))


def _num_states(k: int) -> int:
    return 256 if k == 9 else (64 if k == 7 else 16)


def vstate_lshift(reg: int, k: int, val: int) -> int:
    """Left-shift a K-1 bit register and insert val as the LSB.

    The C decoder stores the newest input at bit 0. The mask keeps only
    the K-1 significant bits.
    """
    mask = {5: 0x0E, 7: 0x3E, 9: 0xFE}.get(k, 0)
    return ((reg << 1) & mask) | val


def _parity(x: int) -> int:
    """Population count mod 2."""
    x ^= x >> 32
    x ^= x >> 16
    x ^= x >> 8
    x ^= x >> 4
    x ^= x >> 2
    x ^= x >> 1
    return x & 1


def gen_state_info(code, reg: int):
    """Populate non-recursive trellis state info.

    For a state (the K-1 bit register), returns:
      val  — the input bit that drove the transition to this state
      out  — the 3 NRZ output symbols (+1/-1) of that transition

    In the C code this is gen_state_info(); the "previous '0' state"
    is computed by shifting the register with a 0 input.
    """
    k, gens = code
    prev = vstate_lshift(reg, k, 0)
    val = (reg >> (k - 2)) & 0x01
    prev |= val << (k - 1)
    out = [(_parity(prev & gens[i]) * 2 - 1) for i in range(3)]
    return val, out


def _build_trellis(code):
    """Compute the output symbols and input value for each state.

    outs[s] holds the 3 NRZ outputs (+1/-1) of the transition into
    state s; vals[s] is the input bit that led to it.
    """
    ns = _num_states(code[0])
    outs = np.zeros((ns, 3), dtype=np.int64)
    vals = np.zeros(ns, dtype=np.uint8)
    for reg in range(ns):
        val, o = gen_state_info(code, reg)
        vals[reg] = val
        outs[reg, :] = o
    return vals, outs


def conv_decode(in_soft: np.ndarray, length: int, k: int, gens):
    """Viterbi decode a tail-biting convolutional code.

    in_soft: 3 * length NRZ soft bits (+1 = bit 1, -1 = bit 0).
    Returns (bits, confidence) where bits is a uint8 array of `length`
    decoded bits and confidence is max−second_max path metric.
    """
    ns = _num_states(k)
    nsteps = length + 2 * TAIL_BITING_EXTRA

    vals, outs = _build_trellis((k, gens))
    # The butterfly pairs destination states (s, s+half); by trellis
    # symmetry both transitions of a pair share the metric of state s
    # (see gen_state_info). So the metric vector is the first half of
    # the output table.
    outs_pair = outs[: ns // 2, :]

    sums = np.zeros(ns, dtype=np.int64)
    paths = np.zeros((nsteps, ns), dtype=np.int8)

    # Normalization interval, matching the C decoder: the path metrics
    # are int16 there, normalized every INT16_MAX/(n*INT8_MAX) - k steps.
    intrvl = 32767 // (3 * 127) - k

    # Tail-biting: the trellis starts TAIL_BITING_EXTRA input bits
    # before the frame start, wrapping to the beginning of the frame.
    seq_idx = (np.arange(length - TAIL_BITING_EXTRA, length)[:, None]
               + np.arange(nsteps)[None, :])
    seq_idx = seq_idx[0] % length

    # numba loop (indexed via precomputed seq_idx)
    _acs_forward_wrapped(sums, outs_pair, in_soft, seq_idx, paths, intrvl)

    # Traceback: find the best final state...
    best = int(np.argmax(sums))
    # Discard the extra tail-biting steps.
    for i in range(nsteps - 1, length + TAIL_BITING_EXTRA - 1, -1):
        path = int(paths[i, best]) + 1
        best = vstate_lshift(best, k, path)
    # Trace back through the data region.
    bits = np.zeros(length, dtype=np.uint8)
    for i in range(length - 1, -1, -1):
        path = int(paths[i + TAIL_BITING_EXTRA, best]) + 1
        bits[i] = vals[best]
        best = vstate_lshift(best, k, path)
    return bits


@njit(cache=True)
def _acs_forward_wrapped(sums, outs, soft, seq_idx, paths, norm_interval):
    ns = sums.shape[0]
    half = ns // 2
    new = np.empty(ns, dtype=np.int64)
    for i in range(seq_idx.shape[0]):
        base = seq_idx[i] * 3
        b0 = soft[base]
        b1 = soft[base + 1]
        b2 = soft[base + 2]
        # The butterfly must read the PREVIOUS step's metrics — write
        # destinations to a scratch array and swap after the sweep, or
        # later butterflies would consume their own outputs (destinations
        # s+half are sources of the pair s+16 in the next step... in
        # fact destinations overlap sources for ANY in-place update).
        for s in range(half):
            m = outs[s, 0] * b0 + outs[s, 1] * b1 + outs[s, 2] * b2
            state0 = sums[2 * s]
            state1 = sums[2 * s + 1]
            sum0 = state0 + m
            sum1 = state1 - m
            sum2 = state0 - m
            sum3 = state1 + m
            if sum0 > sum1:
                paths[i, s] = -1
                new[s] = sum0
            else:
                paths[i, s] = 0
                new[s] = sum1
            if sum2 > sum3:
                paths[i, s + half] = -1
                new[s + half] = sum2
            else:
                paths[i, s + half] = 0
                new[s + half] = sum3
        for j in range(ns):
            sums[j] = new[j]
        if i % norm_interval == 0:
            mn = sums[0]
            for j in range(1, ns):
                if sums[j] < mn:
                    mn = sums[j]
            for j in range(ns):
                sums[j] -= mn


# Public API mirroring the Go wrappers.
def conv_decode_p1(in_soft):
    return conv_decode(in_soft, P1_FRAME_LEN_FM, *CONV_CODE_K7)


def conv_decode_pids(in_soft):
    return conv_decode(in_soft, PIDS_FRAME_LEN, *CONV_CODE_K7)


def conv_decode_p3_p4(in_soft, length):
    return conv_decode(in_soft, length, *CONV_CODE_K7)


def conv_decode_e1(in_soft, length):
    return conv_decode(in_soft, length, *CONV_CODE_E1)


def conv_decode_e2_e3(in_soft, length):
    return conv_decode(in_soft, length, *CONV_CODE_E2_E3)